services/source/updater.c

```diff
@@ -0,0 +1,355 @@
+        status = 0;
+        update.initialized = true;
+        update.start_address = app_info.addr;
+        update.next_address = update.start_address;
+        update.size = app_info.size;
+        break;
+
+    case PROGRAM_BLOCK :
+        if (update.initialized == false) {
+            status = -1;
+            break;
+        }
+        uint16_t size;
+        cnv8_16(&packet->data[IN_DATA_BYTE + 4], &size);
+        uint32_t address;
+        cnv8_32(&packet->data[IN_DATA_BYTE], &address);
+        if (address != (update.next_address)) {
+            error = "Block out of order";
+            status = -1;
+            break;
+        }
+        uint32_t flash_destination = address;
+
+        uint8_t bank = address < 0x00200000 ? 0 : 1;
+        uint8_t *buf = &packet->data[IN_DATA_BYTE + 6];
+
+        oReturnCheck = priv_Fapi_BlockProgram(bank, flash_destination, (unsigned long)buf, size);
+        if (oReturnCheck) {
+            error = "Failed to write to block";
+            status = -1;
+            break;
+        }
+        update.next_address = address + size;
+        break;
+
+    case ERASE_APP:
+        app_info = eeprom_get_app_info();
+        app_info.exists = 0;
+        priv_eeprom_set_app_info(app_info);
+        set_packet_length(packet, sizeof(int8_t) + 1);
+        status = 0;
+        break;
+
+    case VERIFY_APPLICATION_IMAGE:
+        if (verify_application() != true) {
+            status = -1;
+            break;
+        }
+        set_packet_length(packet, sizeof(int8_t) + 1);
+        status = 0;
+        break;
+
+    case VERIFY_GOLDEN_IMAGE:
+        if (verify_golden() != true) {
+            status = -1;
+            break;
+        }
+        set_packet_length(packet, sizeof(int8_t) + 1);
+        status = 0;
+        break;
+
+    default:
+        error = "No such subservice\n";
+        status = -1;
+    }
+    if (status == -1) {
+        set_packet_length(packet, sizeof(int8_t) + 1);
+    }
+    set_packet_length(packet, sizeof(int8_t) + 1);
+    memcpy(&packet->data[STATUS_BYTE], &status, sizeof(int8_t));
+    return SATR_OK;
+}
+
+/**
+ * @brief
+ *      FreeRTOS updater server task
+ * @details
+ *      Accepts incoming updater service packets and executes the application
+ * @param void* param
+ * @return None
+ */
+void updater_service(void *param) {
+
+    csp_socket_t *sock;
+    sock = csp_socket(CSP_SO_RDPREQ); // require RDP connection
+    csp_bind(sock, TC_UPDATER_SERVICE);
+    csp_listen(sock, SERVICE_BACKLOG_LEN);
+    for (;;) {
+        csp_conn_t *conn;
+        csp_packet_t *packet;
+
+        // wait for connection, timeout
+        if ((conn = csp_accept(sock, DELAY_WAIT_TIMEOUT)) == NULL) {
+            /* timeout */
+            continue;
+        }
+
+        while ((packet = csp_read(conn, 50)) != NULL) {
+            if (updater_app(packet) != SATR_OK) {
+                // something went wrong, this shouldn't happen
+                csp_buffer_free(packet);
```

Comment:
If the csp_send is returning a response, don't you want to return both happy and sad responses?

Reply:
I don't know what you mean